agent/config/file_watcher.go

```diff
@@ -0,0 +1,249 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-hclog"
+)
+
+const timeoutDuration = 200 * time.Millisecond
+
+type FileWatcher struct {
+	watcher          *fsnotify.Watcher
+	configFiles      map[string]*watchedFile
+	logger           hclog.Logger
+	reconcileTimeout time.Duration
+	cancel           context.CancelFunc
+	done             chan interface{}
+	stopOnce         sync.Once
+
+	//EventsCh Channel where an event will be emitted when a file change is detected
+	// a call to Start is needed before any event is emitted
+	// after a Call to Stop succeed, the channel will be closed
+	EventsCh chan *FileWatcherEvent
+}
+
+type watchedFile struct {
+	modTime time.Time
+}
+
+type FileWatcherEvent struct {
+	Filename string
+}
+
+//NewFileWatcher create a file watcher that will watch all the files/folders from configFiles
+// if success a FileWatcher will be returned and a nil error
+// otherwise an error and a nil FileWatcher are returned
+func NewFileWatcher(configFiles []string, logger hclog.Logger) (*FileWatcher, error) {
+	ws, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	w := &FileWatcher{
+		watcher:          ws,
+		logger:           logger.Named("file-watcher"),
+		configFiles:      make(map[string]*watchedFile),
+		EventsCh:         make(chan *FileWatcherEvent),
+		reconcileTimeout: timeoutDuration,
+		done:             make(chan interface{}),
+		stopOnce:         sync.Once{},
+	}
+	for _, f := range configFiles {
+		err = w.add(f)
+		if err != nil {
+			return nil, fmt.Errorf("error adding file %q: %w", f, err)
+		}
+	}
+
+	return w, nil
+}
+
+// Start start a file watcher, with a copy of the passed context.
+// calling Start multiple times is a noop
+func (w *FileWatcher) Start(ctx context.Context) {
+	if w.cancel == nil {
+		cancelCtx, cancel := context.WithCancel(ctx)
+		w.cancel = cancel
+		go w.watch(cancelCtx)
+	}
+}
+
+// Stop the file watcher
+// calling Stop multiple times is a noop
```

Comment:
Also might be useful to say "Stop must only be called after Start" since the sequence of `Stop, Start, Stop` would leave the goroutine running.